### PR TITLE
Feature: new method of DFT+U calculation, add parametor "onsite_radius" for projections modulation.

### LINKED ISSUE #3588: DFT+U: develop a new DFT+U method with variable projectors for on-site occupations.
### Background

The theoretical formula is from https://dptechnology.feishu.cn/wiki/HgspwrZu7i8W5DkyokBcpugtnvb

The test branch is [dftu_new](https://github.com/deepmodeling/abacus-develop/tree/dftu_new)

### PR DESCRIPTION
**Pull request by dyzheng**
### Reminder
- [ ] Have you linked an issue with this pull request?
- [ ] Have you noticed possible changes of behavior below or in the linked issue?
- [ ] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Linked Issue
Fix #3588 

### What's changed?
- I have implemented a new DFT+U method for better SCF convergence than method before.
- New parameter "onsite_radius" can choose the radius of projections.
- "mixing_restart" with "mixing_dmr" can approve the SCF convergence for DFT+U calculation.

I will add Unit Tests in future PRs. In this PR, only integrate tests is covered.

### Any changes of core modules? (ignore if not applicable)
- I have added a new Operator DFTUNew in operator_lcao/ directory.

### CODE REVIEW

File: docs/advanced/input_files/input-main.md [hunk 1]
```diff
@@ -2586,6 +2588,14 @@ These variables are used to control DFT+U correlated parameters
 
 - **Default**: 0.0
 
+### onsite_radius
```
Comment (WHUweiqingzhou) [inline docs/advanced/input_files/input-main.md]: How do users use old DFT+U method?

File: docs/advanced/input_files/input-main.md [hunk 2]
```diff
@@ -274,9 +274,11 @@
     - [dft\_plus\_u](#dft_plus_u)
     - [orbital\_corr](#orbital_corr)
     - [hubbard\_u](#hubbard_u)
+    - [onsite\_radius](#onsite_radius)
```
Comment (WHUweiqingzhou) [inline docs/advanced/input_files/input-main.md]: Please delete one.

File: docs/advanced/input_files/input-main.md [hunk 3]
```diff
@@ -2612,6 +2622,43 @@ These variables are used to control DFT+U correlated parameters
 
 - **Default**: 0
 
+### onsite_radius
```
Comment (WHUweiqingzhou) [inline docs/advanced/input_files/input-main.md:2617]: Just keep one and delete the other.

File: source/module_base/test/CMakeLists.txt [hunk 4]
```diff
@@ -9,11 +9,12 @@ AddTest(
   TARGET base_atom_in
   SOURCES atom_in_test.cpp
 )
-AddTest(
-  TARGET base_blacs_connector
-  LIBS ${math_libs}
-  SOURCES blacs_connector_test.cpp
-)
+# Parallel test related to MPI should not be included in the base test set
```
Comment (WHUweiqingzhou) [inline source/module_base/test/CMakeLists.txt]: delete useless part.

File: source/module_basis/module_nao/projgen.cpp [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+                overlap_alpha_chi_left = overlap_alpha_chi;
+                sigma_right = sigma_right + (sigma_right - sigma_left) * 0.5;
+                smooth_sigma(sigma_right);
+                overlap_alpha_chi_right = overlap_dalpha_dchi();
+            }
+        }
+    }
+}
+
+/*int main() {
```
Comment (WHUweiqingzhou) [inline source/module_basis/module_nao/projgen.cpp]: delete this part.

File: source/module_basis/module_nao/test/CMakeLists.txt [hunk 6]
```diff
@@ -14,6 +14,7 @@ AddTest(
     ../atomic_radials.cpp
     ../radial_set.cpp
     ../numerical_radial.cpp
+    ../projgen.cpp
```
Comment (WHUweiqingzhou) [inline source/module_basis/module_nao/test/CMakeLists.txt:17]: Please add UTs.

File: source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_new.cpp [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+            if (this->ucell->cal_dtau(iat0, iat1, R_index1).norm() * this->ucell->lat0
+                < orb.Phi[T1].getRcut() + GlobalV::onsite_radius)
+            {
+                is_adj[ad1] = true;
+            }
+        }
+        filter_adjs(is_adj, adjs);
+        this->adjs_all.push_back(adjs);
+        
+        /*for (int ad1 = 0; ad1 < adjs.adj_num + 1; ++ad1)
```
Comment (WHUweiqingzhou) [inline source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/dftu_new.cpp]: delete it.

File: tests/integrate/260_NO_DJ_PK_PU_FM/INPUT [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+cal_force     1
+cal_stress    1
+#force_thr_ev    0.01
+#relax_method    cg
+#relax_bfgs_init    0.5
+
+mixing_type    broyden
+mixing_beta    0.4
+mixing_restart 5
```
Comment (WHUweiqingzhou) [inline tests/integrate/260_NO_DJ_PK_PU_FM/INPUT]: use new `mixing_restart`

File: tests/integrate/260_NO_DJ_PK_PU_S1/INPUT [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+cal_force     1
+cal_stress    1
+#force_thr_ev    0.01
+#relax_method    cg
+#relax_bfgs_init    0.5
+
+mixing_type    broyden
+mixing_beta    0.4
+mixing_restart 5
```
Comment (WHUweiqingzhou) [inline tests/integrate/260_NO_DJ_PK_PU_S1/INPUT]: use new `mixing_restart`

File: tests/integrate/260_NO_DJ_PK_PU_SO/INPUT [hunk 10]
```diff
@@ -14,27 +14,28 @@ out_chg    0
...
 
-#cal_force     1
-#cal_stress    1
 #force_thr_ev    0.01
 #relax_method    cg
 #relax_bfgs_init    0.5
 
 mixing_type    broyden
 mixing_beta    0.4
+mixing_restart 5
```
Comment (WHUweiqingzhou) [inline tests/integrate/260_NO_DJ_PK_PU_SO/INPUT]: use new `mixing_restart`

File: docs/advanced/input_files/input-main.md (hunk 1 above)
Comment (dyzheng) [reply docs/advanced/input_files/input-main.md]: set dft_plus_u = 2

File: source/module_io/input.h [hunk 11]
```diff
@@ -485,7 +486,7 @@ class Input
     //==========================================================
     //    DFT+U       Xin Qu added on 2020-10-29
     //==========================================================
-    bool dft_plus_u;             ///< true:DFT+U correction; false: standard DFT calculation(default)
+    int dft_plus_u;             ///< true:DFT+U correction; false: standard DFT calculation(default)
```
Comment (WHUweiqingzhou) [inline source/module_io/input.h]: modify the corresponding description.

File: source/module_io/input.cpp [hunk 12]
```diff
@@ -552,7 +553,7 @@ void Input::Default(void)
     //==========================================================
     //    DFT+U     Xin Qu added on 2020-10-29
     //==========================================================
-    dft_plus_u = false; // 1:DFT+U correction; 0: standard DFT calcullation
+    dft_plus_u = 0; // 1:DFT+U correction; 0: standard DFT calcullation
```
Comment (WHUweiqingzhou) [inline source/module_io/input.cpp]: modify the corresponding comment.

File: source/module_base/test/CMakeLists.txt (hunk 4 above)
Comment (dyzheng) [reply source/module_base/test/CMakeLists.txt]: I will recover the original code and raise a new issue for this UT.

**Comment by dyzheng** [comment]
I have optimized the performance of DFT+U code in dftu_new branch, please review again.

**Review by WHUweiqingzhou — APPROVED** [review]
Congratulation!
ABACUS has suffered from bad feedbacks about DFTU from the side of users for a long time.
This PR will provide a refresh new DFTU module, which is expected to be much better than the old one. 
We have made heavy tests to make sure the new one has a decent performance about convergence, magnetic moment, band gap, and so on.

**Pull request merged**